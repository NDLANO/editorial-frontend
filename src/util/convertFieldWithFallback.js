/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

export function convertFieldWithFallback(element, field, fallback, language) {
  if (language) {
    return element[field] && element[field].language === language
      ? element[field][field]
      : fallback;
  }

  return typeof element[field] === 'object' && element[field] !== null
    ? element[field][field]
    : fallback;
}
