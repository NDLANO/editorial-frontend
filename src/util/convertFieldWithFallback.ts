/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

type LangObj<F extends string, R> = { [key in F]: R } & { language: string };
type Element<F extends string, R> = { [key in F]?: R | LangObj<F, R> };

export function convertFieldWithFallback<Key extends string, ReturnType = string>(
  element: Element<Key, ReturnType>,
  fieldName: Key,
  fallback: ReturnType,
  language?: string,
): ReturnType {
  const elementField = element[fieldName];
  if (elementField === null || elementField === undefined) {
    return fallback;
  }

  if (typeof elementField !== "object" || Array.isArray(elementField)) {
    return elementField as ReturnType;
  }

  const langObj = elementField as LangObj<Key, ReturnType>;

  if (language) {
    return langObj.language === language ? langObj[fieldName] : fallback;
  }

  return langObj[fieldName];
}
