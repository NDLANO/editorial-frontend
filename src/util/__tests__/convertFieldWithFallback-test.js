/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { convertFieldWithFallback } from '../convertFieldWithFallback';

const article = {
  id: '2',
  created: '2014-12-24T10:44:06Z',
  title: { title: 'Tester', language: 'nb' },
  metaDescription: { metaDescription: 'Beskrivelse', language: 'nb' },
  visualElement: { content: '<embed />', language: 'nb' },
};

test('convertFieldWithFallback convert field title to string', () => {
  const obj = convertFieldWithFallback(article, 'title', '');
  expect(obj).toMatchSnapshot();
});

test('convertFieldWithFallback convert field with fallback', () => {
  const obj = convertFieldWithFallback(article, 'unkownfield', []);
  expect(obj).toMatchSnapshot();
});

test('convertFieldWithFallback convert field visual element', () => {
  const obj = convertFieldWithFallback(article, 'visualElement', []);
  expect(obj).toMatchSnapshot();
});
