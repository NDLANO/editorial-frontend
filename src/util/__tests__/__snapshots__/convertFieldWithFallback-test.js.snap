// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertFieldWithFallback convert field title to string 1`] = `"Tester"`;

exports[`convertFieldWithFallback convert field visual element 1`] = `"<embed />"`;

exports[`convertFieldWithFallback convert field with fallback 1`] = `Array []`;
